File: src/main/scala/committee/nova/skillful/player/capabilities/ISkills.scala
package committee.nova.skillful.player.capabilities

import committee.nova.skillful.api.ISkill
import committee.nova.skillful.skills.SkillInstance
import committee.nova.skillful.storage.SkillfulStorage
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.ResourceLocation

import scala.collection.mutable

trait ISkills {
  def getSkills: mutable.HashSet[SkillInstance]

  def getSkillInfos: mutable.HashSet[SkillInfo]

  def getSkillInfo(player: EntityPlayer, id: ResourceLocation): SkillInfo = getSkillInfo(player, SkillfulStorage.getSkill(id))

  def getSkillInfo(player: EntityPlayer, skill: ISkill): SkillInfo

  def getSkill(skill: ISkill): SkillInstance

  def getSkill(id: ResourceLocation): SkillInstance = getSkill(SkillfulStorage.getSkill(id))
}

File: src/main/java/committee/nova/skillful/common/cap/skill/ISkills.java
package committee.nova.skillful.common.cap.skill;

import committee.nova.skillful.common.skill.DummySkillInstance;
import committee.nova.skillful.common.skill.ISkillType;
import committee.nova.skillful.common.skill.SkillInstance;
import net.minecraft.nbt.ListTag;
import net.minecraft.resources.ResourceLocation;
import net.minecraftforge.common.util.INBTSerializable;

import java.util.Collection;
import java.util.Optional;

public interface ISkills extends INBTSerializable<ListTag> {
    Collection<SkillInstance> getSkills();

    boolean hasSkill(ISkillType skill);

    Optional<SkillInstance> getSkill(ISkillType skill);

    SkillInstance getOrCreateSkill(ISkillType skill);

    Optional<DummySkillInstance> getUnregisteredSkill(ResourceLocation id);

    void removeUnregisteredSkill(ResourceLocation id);
}
